### ISSUE TITLE
Interaction with --gc-sections ?

### ISSUE THREAD

**Issue by dascandy** [doctest/doctest#8]
When compiling a testcase with -ffunction-sections and -fdata-sections, and linking with --gc-sections (or alternatively, -flto entirely), do the test cases get removed if there is no runner in the executable, or if it is not linked in?

**Comment by onqtam** [comment]
The ```DOCTEST_CONFIG_DISABLE``` identifier should be defined everywhere and not just where the test runner gets instantiated (when removing everything testing-related from the binary).

When it is defined - all test case macros turn into a template that never gets instantiated - and thus shouldn't reside in the binary - even in debug builds - see [here](https://github.com/onqtam/doctest/blob/master/doctest/doctest.h#L925).

Does this answer your question?

**Comment by onqtam** [comment]
The ```DOCTEST_CONFIG_DISABLE``` identifier should be defined everywhere - and not just where the test runner gets implemented.

When it is defined all test cases are no longer registered - and are made into an uninstantiated template - so they shouldn't be in the binary even in debug builds.

Does this answer your question?

**Comment by dascandy** [comment]
Not really. I would want to compile all projects once, and then at link time either link in a test runner or a normal main entrypoint. In the former case, it'll run all tests and exit, in the second case no reference to the tests should exist.

That means you can compile stuff once and have two output builds, by way that unused code elimination works. That also means your shipping code *is* the same build that you test.

I know you can use macros. That's not the question. Does it work with unused code elimination too, or is there some reason that won't work?

**Comment by onqtam** [comment]
I'll have to investigate this.

**Comment by nscooling** [comment]
You should be able to determine by looking at the map file; but it will also depend on you optimisation settings. The GNU toolchain typically will link-time optimise out unreferenced functions.

**Comment by onqtam** [comment]
I won't get to this soon, but I don't think a binary will be able to link in the first place if a test runner is not supplied - for example the self-registering tests reference functions from the test runner.

But if a test runner is also compiled/linked but is not used (no ```doctest::Context``` gets created/used ever) then there will be a strongly connected component of functions which reference each-other but the main program execution will not be able to get to - in that case in theory link-time optimization or some other flags should be able to remove them altogether from the binary - but I doubt this is the case.

**Comment by Trass3r** [comment]
I tested this with just DOCTEST_CONFIG_IMPLEMENT and an empty main but the tests are still referenced from the global constructors code.

Related to https://github.com/onqtam/doctest/issues/53#issuecomment-658310943.

**Comment by Trass3r** [comment]
I think this might be (almost) doable after all, albeit not in a standard way.
Instead of abusing global ctors for test registration you could place test metadata in a special section and rely on the same-name input section merging typically done by the linker. Registration would then be done by the test main.
The problem is that in the face of gc-sections you'd probably need to use [[gnu::used]] so they don't get discarded too early and then you're back at square 1.

**Comment by mitchgrout** [comment]
I'm gonna call this; without a fundamental change to how test auto-registration works, the only way to avoid having tests in the final binary is by compiling under `DOCTEST_CONFIG_DISABLE`. Because we introduce enough indirection via global-init, the linker's liable to leave stuff in. 

The proposed solution of using `section` is fun, but will cause issues if custom linker-scripts are being used.
